### PR TITLE
Avoid reserializing results into multiprocessing in Work Queue and Task Vine executors

### PR DESCRIPTION
**Pull request by tphung3**
# Description

This PR fixes issue #2908 

Fixes # 2908

## Type of change

Choose which options apply, and delete the ones which do not apply.

- Bug fix (non-breaking change that fixes an issue)

### CODE REVIEW

**Comment by tphung3** [comment]
@benclifford I think this fixes issue 2908.

**Comment by benclifford** [comment]
@tphung3 I just re-enabled the test that demonstrates issue #2908 (from PR #2904) - I think that should pass now with your changes.

File: parsl/executors/workqueue/executor.py [hunk 1]
```diff
@@ -727,11 +727,18 @@ def _collect_work_queue_results(self):
...
+                    except Exception as e:
+                        logger.debug(f'Cannot load result from result file {task_report.result_file}. Exception: {e}')
+                        future.set_exception(WorkQueueTaskFailure('Cannot load result from result file', None))
+                    else:
+                        future.set_result(result)
                 else:
                     # If there are no results, then the task failed according to one of
                     # work queue modes, such as resource exhaustion.
-                    future.set_exception(WorkQueueTaskFailure(task_report.reason, task_report.result))
+                    future.set_exception(WorkQueueTaskFailure(task_report.reason, None))
```
[the remark this answers is not in the document]
Comment (tphung3) [reply parsl/executors/workqueue/executor.py:741]: Sure I think good error messages are always beneficial. Do you want to merge #2910 when it is ready, I'll update this PR then.

[the remark this answers is not in the document]
Comment (benclifford) [reply parsl/executors/workqueue/executor.py:741]: If you are happy with #2910 I'll merge it now, and then you can work on the conflicts that will arise with this PR

[the remark this answers is not in the document]
Comment (benclifford) [reply parsl/executors/workqueue/executor.py:741]: @tphung3 that's merged now and conflicts ready for you to resolve

**Comment by tphung3** [comment]
The CI failure doesn't seem to have anything to do with the proposed changes. @benclifford Can this be a transient error?

**Comment by benclifford** [comment]
@tphung3 yeah there is a hang that happens fairly often in local tests that I don't think is easy to fix because it's coming from some awkward architectural choices years ago -- right now the answer is "suck it up and hit retry", and I hope to address what I think is the problem in the next year.

I'll hit retry.

File: parsl/executors/workqueue/executor.py [hunk 2]
```diff
@@ -729,14 +729,25 @@ def _collect_work_queue_results(self):
...
+                # If result_received, then there's a result file. The object inside the file
+                # may be a valid result or an exception caused within the function invocation.
+                # Otherwise there's no result file, implying errors from WorkQueue.
                 if task_report.result_received:
-                    future.set_result(task_report.result)
+                    try:
+                        with open(task_report.result_file, 'rb') as f_in:
+                            result = deserialize(f_in.read())
+                    except Exception as e:
+                        logger.debug(f'Cannot load result from result file {task_report.result_file}. Exception: {e}')
```
Comment (benclifford) [inline parsl/executors/workqueue/executor.py:740]: probably log this as an error not a debug?

File: parsl/executors/workqueue/executor.py [hunk 3]
```diff
@@ -729,14 +729,25 @@ def _collect_work_queue_results(self):
...
+                            future.set_exception(result)
+                        else:
+                            future.set_result(result)
                 else:
                     # If there are no results, then the task failed according to one of
                     # work queue modes, such as resource exhaustion.
-                    ex = WorkQueueTaskFailure(task_report.reason, task_report.result)
-                    if task_report.result is not None:
-                        ex.__cause__ = task_report.result
+                    ex = WorkQueueTaskFailure(task_report.reason, Exception(task_report.reason))
```
Comment (benclifford) [inline parsl/executors/workqueue/executor.py:750]: I think task_report.result is an exception object that has its own stack trace and more detailed reason for failure:

I feel like that exception object should be chained on using `__cause__` to give more detailed stack traces that preserve the information in the original exception, based on the principle that this code (like most code) is not in the right place to discard exception information.

Likewise for a bunch of other places where there's a task failure exception in an exception context (eg line 741, and the equivalents in taskvine)

Comment (tphung3) [reply parsl/executors/workqueue/executor.py:750]: In the line you referred to there's no result file transferred back meaning that it's an internal WQ error. The exception chaining happens in line 743 where a result file comes back and it contains an exception. The executor sets the original exception then. I'll wrap the exception into a WQ Task Failure to make it clearer.

Comment (benclifford) [reply parsl/executors/workqueue/executor.py:750]: right, but if there is *ever* an exception *anywhere*, I don't want it ever discarded apart from at the very outer fringe of parsl

Comment (tphung3) [reply parsl/executors/workqueue/executor.py]: I think the new change adds function chaining whenever it is possible for WQ. Am I understanding you correctly @benclifford ?

Comment (benclifford) [reply parsl/executors/workqueue/executor.py]: @tphung3 yes looks good for Work Queue - maybe Task Vine should have the same behaviour, though? This bit of the code is basically the same in both implementations. Other than that, this is ready to merge I think.

**Comment by benclifford** [comment]
@tphung3 a note on force-pushing to a collaborative branch: looks like your force push here undid the test re-enabling I made, which I didn't notice https://github.com/Parsl/parsl/compare/321901aba999465681b0fe4ca5cd1437a38bc9cb..dbb25270b3dc0edec3b37b8ea16deecebfa6b4b8

**Comment by tphung3** [comment]
My bad. Should have pulled the master merges first.

**Pull request merged**